### DISCUSSION TITLE
Smart Meter Interface commands are not accepted in self-compiled binary on ESP32

### DISCUSSION THREAD

**Discussion by kobuki** [arendst/Tasmota#22327] [Support request] [answered]
I compiled tasmota32 for a generic board to use with the new electricity meter with P1 port installed at our home. Yet it doesn't seem to understand the meter descriptor syntax specified in the [docs](https://tasmota.github.io/docs/Smart-Meter-Interface/). I'm not sure if it's a bug or I'm missing something very simple.

My overrides are:

```
#ifndef USE_SCRIPT
#define USE_SCRIPT
#endif
#ifndef USE_SML_M
#define USE_SML_M
#endif
#ifdef USE_RULES
#undef USE_RULES
#endif

#define SML_BSIZ 1060
#define MAX_METERS 1
#define TMSBSIZ 2048
#define USE_SML_SCRIPT_CMD
#define SML_REPLACE_VARS
```

Then I started with a simple script found on the [docs page](https://tasmota.github.io/docs/Smart-Meter-Interface/):

```
>D
>B
->sensor53 r
>M 1
+1,3,m,0,9600,ABB,1,10,01035B00,01035B02,01035B04,01035B14,01035B16,01035B18,01035B1A,r010350080004,r010350000004,r010350040004
1,010304UUuuUUuu@i0:10,Voltage L1-N,V,Voltage_L1,1
1,010304UUuuUUuu@i1:10,Voltage L2-N,V,Voltage_L2,1
1,010304UUuuUUuu@i2:10,Voltage L3-N,V,Voltage_L3,1
1,010304SSssSSss@i3:100,Active power Total,W,Active_power_Total,2
1,010304SSssSSss@i4:100,Active power L1-N,W,Active_power_L1,2
1,010304SSssSSss@i5:100,Active power L2-N,W,Active_power_L2,2
1,010304SSssSSss@i6:100,Active power L3-N,W,Active_power_L3,2
1,010308xxxxxxxxSSssSSss@i7:100,Real energy,kWh,Real_energy,2
1,010308xxxxxxxxUUuuUUuu@i8:100,Real energy consumed,kWh,Real_energy_consumed,2
1,010308xxxxxxxxUUuuUUuu@i9:100,Real energy delivered,kWh,Real_energy_delivered,2
#
```

But I get this:

```
BRY: Exception> 'syntax_error' - input:1: unexpected symbol near '>'
```

Any `>` command I try throws this message. I checked the binary for strings and it's indeed the one I compiled and the default Tasmota name is "ESP32-DevKit" so I think the compiled binary is OK.

What am I missing? I'm supposed to enter the above descriptor/script in the Berry console, right? The docs doesn't say a single word about that, nor about how I'm supposed to permanently save the config I create.

**Comment by sfromis** [comment, marked as answer]
Nope, not right. You should not use the "Berry Scripting console" as that script has nothing to do with the Berry language, instead use "Edit Script".

**Comment by kobuki** [comment]
Huh. Indeed - no idea why I used the console. However, part of the confusion must have stemmed from the fact stated on the docs page in the first paragraph: "Based on Tasmota's [scripting language](https://tasmota.github.io/docs/Scripting-Language/)." So it's not really true that the 2 are not related. Admittedly, the syntax is different but the Smart Meter Interface uses the same layout as any Berry script. I should have read the scripting docs more carefully. Thanks for the heads-up.

**Reply by sfromis** [reply]
The Berry language and the old scripter feature are totally unrelated. And the Smart Meter Interface is not at all using "same layout" as Berry, but instead can be part of scripter code.

Wording is a bit confusing, as the word "scripting" is basically an adjective commonly attributed to languages meant to be easy to change and run, without prioritizing max speed. Hence, Berry can be described as a "scripting language", while the scripter feature (aka "scripting language") is an older feature from long before ESP32 and Berry were available for Tasmota.

As the docs page you link also talks about "the scripter", this is (sort of) the name, which is also part of the file name `xdrv_10_scripter.ino` where it is implemented, which also consistently talks about "scripter" instead of "scripting language".

Yeah, that's a bit of a mess.

**Reply by kobuki** [reply]
Ah, OK, you're right. I never used any of the 2 extensively (a bit on ESP8266 of the old scripting), and I did confuse them. You learn something every day.
